lib/util/paging-iterator.js

```diff
@@ -0,0 +1,190 @@
+// Requirements
+// -----------------------------------------------------------------------------
+
+var querystring = require('querystring'),
+	Promise = require('bluebird'),
+	PromiseQueue = require('promise-queue'),
+	errors = require('./errors');
+
+PromiseQueue.configure(Promise);
+
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+const PAGING_MODES = Object.freeze({
+	MARKER: 'marker',
+	OFFSET: 'offset'
+});
+
+// -----------------------------------------------------------------------------
+// Public
+// -----------------------------------------------------------------------------
+
+class PagingIterator {
+
+	/**
+	 * Determine if a response is iterable
+	 * @param {Object} response - The API response
+	 * @returns {bool} Whether the response is iterable
+	 */
+	static isIterable(response) {
+		return response.request.method === 'GET' && Array.isArray(response.body.entries);
+	}
+
+	/**
+	 * Asynchronous iterator for paged collections
+	 * @constructor
+	 * @param {Object} response - The original API response
+	 * @param {BoxClient} client - An API client to make further requests
+	 * @returns {void}
+	 * @throws {Error} Will throw when collection cannot be paged
+	 */
+	constructor(response, client) {
+
+		var data = response.body;
+
+		if (!Array.isArray(data.entries)) {
+			throw new Error('Cannot create paging iterator for non-paged response!');
+		}
+
+		if (Number.isSafeInteger(data.offset)) {
+			this.nextField = PAGING_MODES.OFFSET;
+			this.nextValue = data.offset;
+		} else if (typeof data.next_marker !== 'undefined') {
+			this.nextField = PAGING_MODES.MARKER;
+			this.nextValue = data.next_marker;
+		} else {
+			throw new Error('Unable to determine paging strategy for response!');
+		}
+
+		this.limit = data.limit || data.entries.length;
+		this.done = false;
+
+		var queryIndex = response.request.href.indexOf('?');
+		var href = queryIndex > 0 ? response.request.href.substring(0, queryIndex) : response.request.href;
+		this.options = {
+			headers: response.request.headers,
+			qs: querystring.parse(response.request.uri.query)
+		};
+
+		// qs.parse() makes everything a string, ensure the limit is numeric
+		if (this.options.qs.limit) {
+			this.options.qs.limit = parseInt(this.options.qs.limit, 10);
+		}
+
+		delete this.options.headers.Authorization;
+		this.fetch = client.get.bind(client, href);
+		this.nextChunk = response.body.entries;
+		this.queue = new PromiseQueue(1, Infinity);
+		this._updatePaging(response);
+	}
+
+	/**
+	 * Update the paging parameters for the iterator
+	 * @private
+	 * @param {Object} response - The latest API response
+	 * @returns {void}
+	 */
+	_updatePaging(response) {
+
+		var data = response.body;
+
+		if (this.nextField === PAGING_MODES.OFFSET) {
+
+			this.nextValue += this.limit;
```

Comment:
This already uses the limit returned by the API (or the length or the `entries` array if the API didn't return a `limit` field; see the constructor for details